### PR TITLE
Refactor contest results queries

### PR DESCRIPTION
**Pull request by MTsatsarov**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/819

### CODE REVIEW

File: Open Judge System/Data/OJS.Data/OjsDbContext.cs [hunk 1]
```diff
@@ -234,5 +252,37 @@ private int SaveChangesWithTracingDbExceptions()
...
+            }
+            catch (DbUpdateException ex)
+            {
+                Exception currentException = ex;
+                while (currentException != null)
+                {
+                    Trace.TraceError(currentException.Message);
+                    currentException = currentException.InnerException;
+                }
+
```
Comment (gogo4ds) [inline Open Judge System/Data/OJS.Data/OjsDbContext.cs:270]: You can extract method `HandleDbUpdateException` and use it both sync and async versions.

File: Open Judge System/Data/OJS.Data/OjsDbContext.cs [hunk 2]
```diff
@@ -234,5 +252,37 @@ private int SaveChangesWithTracingDbExceptions()
...
+            catch (DbEntityValidationException ex)
+            {
+                foreach (var validationErrors in ex.EntityValidationErrors)
+                {
+                    foreach (var validationError in validationErrors.ValidationErrors)
+                    {
+                        Trace.TraceError(
+                            $"Property: {validationError.PropertyName}{Environment.NewLine} Error: {validationError.ErrorMessage}");
+                    }
+                }
```
Comment (gogo4ds) [inline Open Judge System/Data/OJS.Data/OjsDbContext.cs:282]: Same here => `HandleDbEntityValidationException(DbEntityValidationException ex)`

File: Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs [hunk 3]
```diff
@@ -487,41 +509,77 @@ public ActionResult StatsChart(int contestId)
 
             var participants = this.participantsData
                 .GetAllByContestAndIsOfficial(contest.Id, official)
-                .AsNoTracking();
+                .ToList()
+                .OrderByDescending(p => p.TotalScoreSnapshot);
```
Comment (gogo4ds) [inline Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs:513]: Try not to get all the participants at once in memory, where not necessary (results pages).

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Data/OJS.Data/OjsDbContext.cs (hunk 1 above)
Comment (MTsatsarov) [reply Open Judge System/Data/OJS.Data/OjsDbContext.cs]: Done

File: Open Judge System/Data/OJS.Data/OjsDbContext.cs (hunk 2 above)
Comment (MTsatsarov) [reply Open Judge System/Data/OJS.Data/OjsDbContext.cs]: Done

File: Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs (hunk 3 above)
Comment (MTsatsarov) [reply Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs]: Done

File: Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs [hunk 4]
```diff
@@ -507,12 +507,32 @@ public ActionResult StatsChart(int contestId)
...
+                    ? OfficialResultsPageSize
+                    : NotOfficialResultsPageSize;
+            }
+
             var participants = this.participantsData
                 .GetAllByContestAndIsOfficial(contest.Id, official)
-                .ToList()
-                .OrderByDescending(p => p.TotalScoreSnapshot);
+                .OrderByDescending(p => p.TotalScoreSnapshot)
+                .Skip(participantsToSkip)
```
Comment (gogo4ds) [inline Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs:528]: ThenBy(p => p.TotalScoreSnapshotModifiedOn). We want the users who completed the contest first, to appear first in the results.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs (hunk 4 above)
Comment (MTsatsarov) [reply Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs]: Done

**Review by gogo4ds — APPROVED** [review]

**Pull request merged**